### ISSUE TITLE
Error when using dirac-dms-directory-sync

### ISSUE THREAD

**Issue by HealthyPear** [DIRACGrid/DIRAC#5775]
- I am using DIRAC 7.3.12 installed via pip in a conda environment

I launch the command as 
`dirac-dms-directory-sync SOURCE DESTINATION`

where `SOURCE` is correctly defined, as I reach it using `dirac-dms-filecatalog-cli` and I can download files or list of files from it.

The error I get is the following,

```python
Traceback (most recent call last):
  File "/Users/michele/Applications/miniconda3/envs/protopipe-CTADIRAC-dev/bin/dirac-dms-directory-sync", line 8, in <module>
    sys.exit(main())
  File "/Users/michele/Applications/miniconda3/envs/protopipe-CTADIRAC-dev/lib/python3.8/site-packages/DIRAC/Core/Utilities/DIRACScript.py", line 80, in __call__
    return entrypointFunc._func()
  File "/Users/michele/Applications/miniconda3/envs/protopipe-CTADIRAC-dev/lib/python3.8/site-packages/DIRAC/DataManagementSystem/scripts/dirac_dms_directory_sync.py", line 515, in main
    returnValue = run(args, sync, parallel)
  File "/Users/michele/Applications/miniconda3/envs/protopipe-CTADIRAC-dev/lib/python3.8/site-packages/DIRAC/DataManagementSystem/scripts/dirac_dms_directory_sync.py", line 508, in run
    res = syncDestinations(upload, source_dir, dest_dir, storage, delete, nthreads)
  File "/Users/michele/Applications/miniconda3/envs/protopipe-CTADIRAC-dev/lib/python3.8/site-packages/DIRAC/DataManagementSystem/scripts/dirac_dms_directory_sync.py", line 476, in syncDestinations
    res = doDownload(dm, result, source_dir, dest_dir, delete, nthreads)
  File "/Users/michele/Applications/miniconda3/envs/protopipe-CTADIRAC-dev/lib/python3.8/site-packages/DIRAC/DataManagementSystem/scripts/dirac_dms_directory_sync.py", line 400, in doDownload
    res = runInParallel(
  File "/Users/michele/Applications/miniconda3/envs/protopipe-CTADIRAC-dev/lib/python3.8/site-packages/DIRAC/DataManagementSystem/scripts/dirac_dms_directory_sync.py", line 450, in runInParallel
    pro.start()
  File "/Users/michele/Applications/miniconda3/envs/protopipe-CTADIRAC-dev/lib/python3.8/multiprocessing/process.py", line 121, in start
    self._popen = self._Popen(self)
  File "/Users/michele/Applications/miniconda3/envs/protopipe-CTADIRAC-dev/lib/python3.8/multiprocessing/context.py", line 224, in _Popen
    return _default_context.get_context().Process._Popen(process_obj)
  File "/Users/michele/Applications/miniconda3/envs/protopipe-CTADIRAC-dev/lib/python3.8/multiprocessing/context.py", line 284, in _Popen
    return Popen(process_obj)
  File "/Users/michele/Applications/miniconda3/envs/protopipe-CTADIRAC-dev/lib/python3.8/multiprocessing/popen_spawn_posix.py", line 32, in __init__
    super().__init__(process_obj)
  File "/Users/michele/Applications/miniconda3/envs/protopipe-CTADIRAC-dev/lib/python3.8/multiprocessing/popen_fork.py", line 19, in __init__
    self._launch(process_obj)
  File "/Users/michele/Applications/miniconda3/envs/protopipe-CTADIRAC-dev/lib/python3.8/multiprocessing/popen_spawn_posix.py", line 47, in _launch
    reduction.dump(process_obj, fp)
  File "/Users/michele/Applications/miniconda3/envs/protopipe-CTADIRAC-dev/lib/python3.8/multiprocessing/reduction.py", line 60, in dump
    ForkingPickler(file, protocol).dump(obj)
AttributeError: Can't pickle local object 'main.<locals>.downloadListOfFiles' 
```

**Comment by fstagni** [comment]
Thanks for reporting. In order to exclude setup errors, would you please trying installing a DIRAC client using DIRACOS? Follow instructions in https://dirac.readthedocs.io/en/latest/UserGuide/GettingStarted/InstallingClient/index.html
Then try again the above command (with DIRACOS we using python 3.9).

**Comment by chrisburr** [comment]
I know what is causing this so I'll take care of fixing it.

(The bulk conversion of scripts to use `def main(): ... if __name__ == "__main__": main()` wasn't done carefully enough in v7r2 to handle this case)

**Comment by HealthyPear** [comment]
Thanks @chrisburr, I'll test it again when this will be released then.

Meanwhile, what would be the minimal downgrade required to go back to a working version?

**Comment by chrisburr** [comment]
I don't have a good suggestion of what to download to as all Python 3-compatible releases will contain this bug.

I think you can test the fix by running these two commands:
```bash
# Backup the dirac_dms_directory_sync.py file
cp /Users/michele/Applications/miniconda3/envs/protopipe-CTADIRAC-dev/lib/python3.8/site-packages/DIRAC/DataManagementSystem/scripts/dirac_dms_directory_sync.py /Users/michele/Applications/miniconda3/envs/protopipe-CTADIRAC-dev/lib/python3.8/site-packages/DIRAC/DataManagementSystem/scripts/dirac_dms_directory_sync.py.bak
# Download the fixed version of dirac_dms_directory_sync.py
curl -L https://raw.githubusercontent.com/DIRACGrid/DIRAC/6539e3e9ecd72efc5a28067eb47e2152ff3618b8/src/DIRAC/DataManagementSystem/scripts/dirac_dms_directory_sync.py > /Users/michele/Applications/miniconda3/envs/protopipe-CTADIRAC-dev/lib/python3.8/site-packages/DIRAC/DataManagementSystem/scripts/dirac_dms_directory_sync.py
```

Can you report back if this fixed the problem?

**Comment by HealthyPear** [comment]
with that change I get a different error

```python
Traceback (most recent call last):
  File "/Users/michele/Applications/miniconda3/envs/protopipe-CTADIRAC-dev/bin/dirac-dms-directory-sync", line 8, in <module>
    sys.exit(main())
  File "/Users/michele/Applications/miniconda3/envs/protopipe-CTADIRAC-dev/lib/python3.8/site-packages/DIRAC/Core/Utilities/DIRACScript.py", line 80, in __call__
    return entrypointFunc._func()
  File "/Users/michele/Applications/miniconda3/envs/protopipe-CTADIRAC-dev/lib/python3.8/site-packages/DIRAC/DataManagementSystem/scripts/dirac_dms_directory_sync.py", line 534, in main
    returnValue = run(args, sync, parallel)
  File "/Users/michele/Applications/miniconda3/envs/protopipe-CTADIRAC-dev/lib/python3.8/site-packages/DIRAC/DataManagementSystem/scripts/dirac_dms_directory_sync.py", line 490, in run
    res = syncDestinations(upload, source_dir, dest_dir, storage, delete, nthreads)
  File "/Users/michele/Applications/miniconda3/envs/protopipe-CTADIRAC-dev/lib/python3.8/site-packages/DIRAC/DataManagementSystem/scripts/dirac_dms_directory_sync.py", line 457, in syncDestinations
    res = doDownload(dm, result, source_dir, dest_dir, delete, nthreads)
  File "/Users/michele/Applications/miniconda3/envs/protopipe-CTADIRAC-dev/lib/python3.8/site-packages/DIRAC/DataManagementSystem/scripts/dirac_dms_directory_sync.py", line 375, in doDownload
    res = runInParallel(
  File "/Users/michele/Applications/miniconda3/envs/protopipe-CTADIRAC-dev/lib/python3.8/site-packages/DIRAC/DataManagementSystem/scripts/dirac_dms_directory_sync.py", line 428, in runInParallel
    pro.start()
  File "/Users/michele/Applications/miniconda3/envs/protopipe-CTADIRAC-dev/lib/python3.8/multiprocessing/process.py", line 121, in start
    self._popen = self._Popen(self)
  File "/Users/michele/Applications/miniconda3/envs/protopipe-CTADIRAC-dev/lib/python3.8/multiprocessing/context.py", line 224, in _Popen
    return _default_context.get_context().Process._Popen(process_obj)
  File "/Users/michele/Applications/miniconda3/envs/protopipe-CTADIRAC-dev/lib/python3.8/multiprocessing/context.py", line 284, in _Popen
    return Popen(process_obj)
  File "/Users/michele/Applications/miniconda3/envs/protopipe-CTADIRAC-dev/lib/python3.8/multiprocessing/popen_spawn_posix.py", line 32, in __init__
    super().__init__(process_obj)
  File "/Users/michele/Applications/miniconda3/envs/protopipe-CTADIRAC-dev/lib/python3.8/multiprocessing/popen_fork.py", line 19, in __init__
    self._launch(process_obj)
  File "/Users/michele/Applications/miniconda3/envs/protopipe-CTADIRAC-dev/lib/python3.8/multiprocessing/popen_spawn_posix.py", line 47, in _launch
    reduction.dump(process_obj, fp)
  File "/Users/michele/Applications/miniconda3/envs/protopipe-CTADIRAC-dev/lib/python3.8/multiprocessing/reduction.py", line 60, in dump
    ForkingPickler(file, protocol).dump(obj)
TypeError: cannot pickle '_thread.lock' object
```

**Comment by HealthyPear** [comment]
With the fix provided in  #5786 it seems to work for me now

**Comment by chrisburr** [comment]
Thanks for checking so quickly! 😃
